packages/block-editor/src/components/block-list/block-wrapper.js

```diff
@@ -235,11 +236,36 @@ const elements = [
 	'nav',
 ];
 
+const BlockSaveComponent = forwardRef(
+	( { children, tagName: TagName = 'div', ...props }, wrapper ) => {
+		return (
+			<BlockPropsFilterContext.Consumer>
+				{ ( filter ) => {
+					return (
```

Comment:
I'm running into a really strange problem here. `filter` returns the `BlockContent` function from the `BlockContentProvider` (for `InnerBlocks`)... even though I'm subscribing to `BlockPropsFilterContext`. If I swap the two provider components, then it works, but then `InnerBlocks.Content` stops working. @youknowriad Any idea why context is behaving strangely during serialisation? Been staring at this for a while. 🤔

Reply:
I know that we have a custom serializer implementation and we don't rely on React. That might be the reason. cc @aduth

Reply:
Yeah, it could very well be an issue in the implementation of the custom serializer.

In particular, looking at this code, I can see how it might be problematic for two separate context values to coexist in the same render tree, since the second argument would _replace_ the current context:

https://github.com/WordPress/gutenberg/blob/b7125c05187a82ad93f897c8b028295ecf26bd0a/packages/element/src/serialize.js#L413-L414

It's likely not been discovered since our use of context in serialization is pretty minimal to date.

The solution may be a bit tricky, since we context still needs to be associated 1:1 for a given provider/context pair, and we can't assume anything about the shape of that value (i.e. can't be doing any `Object.assign`).

I'd guess we might want to implement context here as some [Weak]Map which can be looked up by Provider class at the time if/when the Consumer is rendered.

Failing test case:

```js
it( 'renders provided value through multiple context providers', () => {
	const {
		Consumer: FirstConsumer,
		Provider: FirstProvider,
	} = createContext();
	const { Provider: SecondProvider } = createContext();

	const result = renderElement(
		<FirstProvider value="first">
			<SecondProvider value="second">
				<FirstConsumer>
					{ ( context ) => context.value }
				</FirstConsumer>
			</SecondProvider>
		</FirstProvider>
	);

	expect( result ).toBe( 'first' );
} );
```

On the general point of the custom serializer, there's other issues here as well (notably, hooks, see #15873). Even if we fixed the above implementation of serializer provider rendering, I'd guess `useContext` likely still wouldn't work? I'm open to reevaluating this implementation, depending how much a burden it continues to be. For context on why it exists in the first place, see #5897 (#3353).

Reply:
See: #21156